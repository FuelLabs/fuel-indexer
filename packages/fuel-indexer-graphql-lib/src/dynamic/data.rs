use super::prelude::*;

#[derive(Clone, Hash, PartialEq, Eq, Debug, EnumString, strum::Display)]
pub enum DynamicDataType {
    Int,
    String,
    // TODO: Implement remaining types
}

pub type DynamicData = serde_json::Value;

pub type DynamicFieldId = String;

#[derive(Clone, Hash, PartialEq, Eq)]
pub struct DynamicDataField(DynamicFieldId, DynamicDataType);
impl DynamicDataField {
    pub fn new(name: impl Into<String>, data_type: impl Into<DynamicDataType>) -> Self {
        Self(name.into(), data_type.into())
    }

    pub fn id(&self) -> &DynamicFieldId {
        &self.0
    }
    pub fn data_type(&self) -> &DynamicDataType {
        &self.1
    }
}

#[derive(Clone)]
pub struct DynamicDataFieldResolver {
    pub data_field: DynamicDataField,
}
impl DynamicDataFieldResolver {
    pub fn new(data_field: DynamicDataField) -> Self {
        Self { data_field }
    }

    pub fn resolve(self, data: &DynamicData) -> Option<FieldValue> {
        let value = data.get(self.data_field.id()).unwrap();
        match self.data_field.data_type() {
            DynamicDataType::String => {
                let value: String = value.as_str().unwrap().to_string();
                Some(FieldValue::value(value))
            }
            DynamicDataType::Int => {
                let value: i32 = value.as_i64().unwrap() as i32;
                Some(FieldValue::value(value))
            }
        }
    }
}
