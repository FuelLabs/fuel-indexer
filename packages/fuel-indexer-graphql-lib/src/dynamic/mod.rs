pub mod connection;
pub mod data;
pub mod edge;
pub mod loader;
pub mod node;
pub mod paging;
pub mod resolver;

pub use connection::*;
pub use data::*;
pub use edge::*;
pub use loader::*;
pub use node::*;
pub use paging::*;
pub use resolver::*;

pub(self) mod prelude {
    pub use crate::prelude::*;
    pub use crate::spec::*;
    pub use crate::util::*;
    pub use async_graphql::dynamic::*;
    pub use std::collections::HashMap;
    pub use std::hash::Hash;
    pub use std::str::FromStr;
    pub use std::sync::Arc;
    pub use strum::EnumString;
}
