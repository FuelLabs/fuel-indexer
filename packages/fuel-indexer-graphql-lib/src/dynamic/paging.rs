use super::prelude::*;

pub type DynamicCursor = Cursor;

#[derive(Clone, PartialEq, Eq)]
pub struct DynamicPageInfo {
    pub has_next_page: bool,
    pub has_previous_page: bool,
    pub start_cursor: Option<DynamicCursor>,
    pub end_cursor: Option<DynamicCursor>,
}
