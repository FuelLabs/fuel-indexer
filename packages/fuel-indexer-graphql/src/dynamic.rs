use std::collections::{HashMap, HashSet};

use async_graphql::{
    dynamic::{
        Enum, Field, FieldFuture, FieldValue, InputObject, InputValue, Object,
        ResolverContext, Scalar, Schema as DynamicSchema,
        SchemaBuilder as DynamicSchemaBuilder, SchemaError, TypeRef,
    },
    Request,
};
use async_graphql_parser::types::{BaseType, Type};
use async_graphql_value::Name;
use fuel_indexer_database::{queries, IndexerConnectionPool};
use fuel_indexer_schema::db::tables::Schema;
use lazy_static::lazy_static;
use serde_json::Value;

use crate::graphql::{GraphqlError, GraphqlQueryBuilder, GraphqlResult};

lazy_static! {
    /// Scalar types supported by the Fuel indexer. These should always stay up-to-date
    /// with fuel-indexer-schema/src/base.graphql.
    static ref SCALAR_TYPES: HashSet<&'static str> = HashSet::from([
        "Address",
        "AssetId",
        "Blob",
        "BlockHeight",
        "Boolean",
        "Bytes32",
        "Bytes4",
        "Bytes64",
        "Bytes8",
        "Charfield",
        "Color",
        "ContractId",
        "HexString",
        "ID",
        "Identity",
        "Int1",
        "Int16",
        "Int4",
        "Int8",
        "Json",
        "MessageId",
        "Nonce",
        "Salt",
        "Signature",
        "Tai64Timestamp",
        "Timestamp",
        "TxId",
        "UInt1",
        "UInt16",
        "UInt4",
        "UInt8",
    ]);

    /// Scalar types that are represented by a numeric type. This ensures that the
    /// value type provided for a field filter matches the type of the scalar itself.
    static ref NUMERIC_SCALAR_TYPES: HashSet<&'static str> = HashSet::from([
        "Int16",
        "Int4",
        "Int8",
        "Timestamp",
        "Tai64Timestamp",
        "UInt16",
        "UInt4",
        "UInt8",
    ]);

    /// Scalar types that are represented by a string type. This ensures that the
    /// value type provided for a field filter matches the type of the scalar itself.
    static ref STRING_SCALAR_TYPES: HashSet<&'static str> = HashSet::from([
        "Address",
        "AssetId",
        "Blob",
        "Bytes32",
        "Bytes4",
        "Bytes64",
        "Bytes64",
        "Bytes8",
        "Charfield",
        "Color",
        "ContractId",
        "HexString",
        "ID",
        "Identity",
        "Json",
        "MessageId",
        "Nonce",
        "Salt",
        "Signature",
    ]);

    /// Scalar types that can be sorted.
    static ref SORTABLE_SCALAR_TYPES: HashSet<&'static str> = HashSet::from([
        "Address",
        "AssetId",
        "Charfield",
        "Color",
        "ContractId",
        "HexString",
        "ID",
        "Identity",
        "Int16",
        "Int4",
        "Int8",
        "MessageId",
        "Nonce",
        "Salt",
        "Signature",
        "Tai64Timestamp",
        "Timestamp",
        "UInt16",
        "UInt4",
        "UInt8",
    ]);

    /// Entity types that should be ignored when building the dynamic schema,
    /// so that they do not appear in the generated documentation. This is done
    /// to hide internal Fuel indexer entity types.
    static ref IGNORED_ENTITY_TYPES: HashSet<&'static str> =
        HashSet::from(["IndexMetadataEntity", "QueryRoot"]);

    /// Entity fields that should be ignored when building the dynamic schema,
    /// so that they do not appear in the generated documentation. This is done
    /// to hide internal Fuel indexer entity fields.
    static ref IGNORED_ENTITY_FIELD_TYPES: HashSet<&'static str> =
        HashSet::from(["object"]);
}

/// Execute user query and return results.
pub async fn execute_query(
    dynamic_request: Request,
    dynamic_schema: DynamicSchema,
    user_query: String,
    pool: IndexerConnectionPool,
    schema: Schema,
) -> GraphqlResult<Value> {
    // Because the schema types from async-graphql expect each field to be resolved
    // separately, it became untenable to use the .execute() method of the dynamic
    // schema itself to resolve queries. Instead, we set it to only resolve
    // introspection queries and then pass any non-introspection queries to our
    // custom query resolver.
    match dynamic_request.operation_name.as_deref() {
        Some("IntrospectionQuery") | Some("introspectionquery") => {
            let introspection_results = dynamic_schema.execute(dynamic_request).await;
            let data = introspection_results.data.into_json()?;

            Ok(data)
        }
        Some(_) | None => {
            let query =
                GraphqlQueryBuilder::new(&schema, user_query.as_str())?.build()?;

            let queries = query.as_sql(&schema, pool.database_type())?.join(";\n");
            let mut conn = match pool.acquire().await {
                Ok(c) => c,
                Err(e) => return Err(GraphqlError::QueryError(e.to_string())),
            };

            match queries::run_query(&mut conn, queries).await {
                Ok(r) => Ok(r),
                Err(e) => Err(GraphqlError::QueryError(e.to_string())),
            }
        }
    }
}

/// Build a dynamic schema. This allows for introspection, which allows for extensive
/// auto-documentation and code suggestions.
pub fn build_dynamic_schema(schema: &Schema) -> GraphqlResult<DynamicSchema> {
    // Register scalars into dynamic schema so that users are aware of their existence.
    let mut schema_builder: DynamicSchemaBuilder = SCALAR_TYPES.iter().fold(
        DynamicSchema::build("QueryRoot", None, None).introspection_only(),
        |sb, scalar| {
            // These types come pre-included in SchemaBuilder.
            if *scalar == "Boolean" || *scalar == "ID" {
                sb
            } else {
                sb.register(Scalar::new(*scalar))
            }
        },
    );

    let mut input_objects = Vec::new();
    let mut filter_object_list = Vec::new();
    let mut filter_tracker = HashMap::new();
    let mut sort_object_list = Vec::new();
    let mut sorter_tracker = HashMap::new();

    // async-graphql requires a root query object so that the base entity
    // fields can be queried against. This QueryRoot does not appear anywhere
    // in the generated documentation nor is it required for the user to create.
    let mut query_root = Object::new("QueryRoot");

    let sort_enum = Enum::new("SortOrder").item("asc").item("desc");

    for (entity_type, field_map) in &schema.fields {
        if IGNORED_ENTITY_TYPES.contains(&entity_type.as_str()) {
            continue;
        }

        // Input values are stored for each field so that we can create
        // a final input object for filtering and sorting a field.
        let mut filter_input_vals = Vec::new();
        let mut sort_input_vals = Vec::new();

        // Field names will be added to this enum in order to allow
        // for filtering on the column itself, i.e. "has" operator.
        let mut object_field_enum = Enum::new(format!("{entity_type}Fields"));

        for (field_name, field_type) in field_map.clone() {
            if IGNORED_ENTITY_FIELD_TYPES.contains(&field_name.as_str()) {
                continue;
            }

            let (field_filter_input_val, mut field_input_objects, sort_input_val) =
                create_input_values_and_objects_for_field(
                    field_name.clone(),
                    field_type,
                    entity_type.clone(),
                    &sort_enum,
                )?;

            filter_input_vals.push(field_filter_input_val);
            input_objects.append(&mut field_input_objects);

            if let Some(input_val) = sort_input_val {
                sort_input_vals.push(input_val);
            }

            object_field_enum = object_field_enum.item(field_name);
        }

        // Fold corresponding input values into objects for that particular field.
        let filter_object = filter_input_vals
            .into_iter()
            .fold(
                InputObject::new(format!("{entity_type}Filter")),
                |input_obj, input_val| input_obj.field(input_val),
            )
            .field(InputValue::new(
                "has",
                TypeRef::named_nn_list(object_field_enum.type_name()),
            ));

        let sort_object = sort_input_vals.into_iter().fold(
            InputObject::new(format!("{entity_type}Sort")),
            |input_obj, input_val| input_obj.field(input_val),
        );

        // For some reason, async-graphql does not implement the Hash trait on any of the
        // type that we need for dynamic schemas. So we are essentially making a hash table
        // ourselves for the filter and sort objects.
        filter_object_list.push(filter_object);
        filter_tracker.insert(entity_type.to_string(), filter_object_list.len() - 1);

        sort_object_list.push(sort_object);
        sorter_tracker.insert(entity_type.to_string(), sort_object_list.len() - 1);

        // Additionally, because we cannot refer to the object fields directly and
        // associate the field arguments to them, we iterate through the fields a
        // second time and construct the fields for the dynamic schema and add the
        // field arguments as well.
        let mut fields = Vec::new();
        for (field_name, field_type) in field_map {
            if IGNORED_ENTITY_FIELD_TYPES.contains(&field_name.as_str()) {
                continue;
            }

            if let Some(field_def) = Type::new(field_type) {
                let base_field_type = &field_def.base;
                let nullable = field_def.nullable;

                let field_type = match base_field_type {
                    BaseType::Named(type_name) => {
                        if nullable {
                            TypeRef::named(type_name.to_string())
                        } else {
                            TypeRef::named_nn(type_name.to_string())
                        }
                    }
                    BaseType::List(list_type) => {
                        let inner_base_type = list_type.base.to_string();
                        let nullable_inner = list_type.nullable;

                        if nullable && nullable_inner {
                            TypeRef::named_list(inner_base_type)
                        } else if nullable && !nullable_inner {
                            TypeRef::named_nn_list(inner_base_type)
                        } else if !nullable && nullable_inner {
                            TypeRef::named_list_nn(inner_base_type)
                        } else {
                            TypeRef::named_nn_list_nn(inner_base_type)
                        }
                    }
                };

                let field = create_field_with_assoc_args(
                    field_name.to_string(),
                    field_type,
                    base_field_type,
                    &filter_tracker,
                    &filter_object_list,
                    &sorter_tracker,
                    &sort_object_list,
                );

                fields.push(field);
            }
        }

        // Create object using all of the fields that were constructed for the entity
        // and repeat the same process in order to allow for introspection-related
        // functionality at the root query level.
        let obj = fields
            .into_iter()
            .fold(Object::new(entity_type.clone()), |obj, f| obj.field(f));

        // Create field for entity object and add it to root level query object.
        let field = create_field_with_assoc_args(
            entity_type.to_string().to_lowercase(),
            TypeRef::named(obj.type_name()),
            &BaseType::Named(Name::new(obj.type_name())),
            &filter_tracker,
            &filter_object_list,
            &sorter_tracker,
            &sort_object_list,
        );
        if !SCALAR_TYPES.contains(&obj.type_name()) {
            query_root = query_root.field(field);
        }

        schema_builder = schema_builder.register(obj).register(object_field_enum);
    }

    // In order for the schema to successfully use the input objects
    // that make up the filter and sort arguments, the objects have to
    // be registered with the schema.
    for filter_obj in filter_object_list {
        schema_builder = schema_builder.register(filter_obj);
    }

    for sort_obj in sort_object_list {
        schema_builder = schema_builder.register(sort_obj);
    }

    for io in input_objects {
        schema_builder = schema_builder.register(io);
    }

    schema_builder = schema_builder.register(sort_enum);
    schema_builder = schema_builder.register(query_root);

    Ok(schema_builder.finish()?)
}

/// Create input values and objects that are used to build introspection information for a field.
fn create_input_values_and_objects_for_field(
    field_name: String,
    field_type: String,
    entity_type: String,
    sort_enum: &Enum,
) -> GraphqlResult<(InputValue, Vec<InputObject>, Option<InputValue>)> {
    let field_type =
        Type::new(&field_type).ok_or(GraphqlError::DynamicSchemaBuildError(
            SchemaError::from("Could not create type defintion from field type string"),
        ))?;

    match field_type.base {
        BaseType::Named(field_type) => {
            let (field_filter_input_val, field_input_objects) =
                create_filter_val_and_objects_for_field(
                    &field_name,
                    field_type.as_str(),
                    entity_type.as_str(),
                );

            if SORTABLE_SCALAR_TYPES.contains(field_type.as_str()) {
                let sort_input_val =
                    InputValue::new(field_name, TypeRef::named(sort_enum.type_name()));
                return Ok((
                    field_filter_input_val,
                    field_input_objects,
                    Some(sort_input_val),
                ));
            }

            Ok((field_filter_input_val, field_input_objects, None))
        }
        // TODO: Do the same as above, but with list type
        BaseType::List(_) => unimplemented!("List types are not currently supported"),
    }
}

fn create_field_with_assoc_args(
    field_name: String,
    field_type_ref: TypeRef,
    base_field_type: &BaseType,
    filter_tracker: &HashMap<String, usize>,
    filter_object_list: &[InputObject],
    sorter_tracker: &HashMap<String, usize>,
    sort_object_list: &[InputObject],
) -> Field {
    // Because the dynamic schema is set to only resolve introspection
    // queries, we set the resolvers to return a dummy value.
    let mut field =
        Field::new(field_name, field_type_ref, move |_ctx: ResolverContext| {
            return FieldFuture::new(async move { Ok(Some(FieldValue::value(1))) });
        });

    match base_field_type {
        BaseType::Named(field_type) => {
            if !SCALAR_TYPES.contains(field_type.as_str()) {
                if let Some(idx) = filter_tracker.get(&field_type.to_string()) {
                    let object_filter_arg = InputValue::new(
                        "filter",
                        TypeRef::named(filter_object_list[*idx].type_name()),
                    );
                    field = field.argument(object_filter_arg)
                }

                if let Some(idx) = sorter_tracker.get(&field_type.to_string()) {
                    let object_sort_arg = InputValue::new(
                        "order",
                        TypeRef::named(sort_object_list[*idx].type_name()),
                    );
                    field = field.argument(object_sort_arg);
                }

                let offset_arg = InputValue::new("offset", TypeRef::named(TypeRef::INT));

                let limit_arg = InputValue::new("first", TypeRef::named(TypeRef::INT));

                let id_selection_arg =
                    InputValue::new("id", TypeRef::named(TypeRef::STRING));

                field = field
                    .argument(offset_arg)
                    .argument(limit_arg)
                    .argument(id_selection_arg);
            }
        }
        BaseType::List(_) => unimplemented!("List types are not currently supported"),
    }

    field
}

/// Build the filter objects for a particular field. The resultant object
/// will ensure that the correct value type is allowed for the field by
/// passing the input type information in the introspection response.
fn create_filter_val_and_objects_for_field<'a>(
    field_name: &'a str,
    field_type: &'a str,
    obj_name: &'a str,
) -> (InputValue, Vec<InputObject>) {
    let mut input_objs: Vec<InputObject> = Vec::new();

    let filter_arg_type = if NUMERIC_SCALAR_TYPES.contains(field_type) {
        TypeRef::INT
    } else {
        TypeRef::STRING
    };

    // TODO: Add support for logical operators -- https://github.com/FuelLabs/fuel-indexer/issues/917

    let complex_comparison_obj =
        InputObject::new(format!("{obj_name}_{field_name}_ComplexComparisonObject"))
            .field(InputValue::new("min", TypeRef::named_nn(filter_arg_type)))
            .field(InputValue::new("max", TypeRef::named_nn(filter_arg_type)));

    let complete_comparison_obj =
        InputObject::new(format!("{obj_name}{field_name}FilterObject"))
            .field(InputValue::new(
                "between",
                TypeRef::named(complex_comparison_obj.type_name()),
            ))
            .field(InputValue::new("equals", TypeRef::named(filter_arg_type)))
            .field(InputValue::new("gt", TypeRef::named(filter_arg_type)))
            .field(InputValue::new("gte", TypeRef::named(filter_arg_type)))
            .field(InputValue::new("lt", TypeRef::named(filter_arg_type)))
            .field(InputValue::new("lte", TypeRef::named(filter_arg_type)))
            .field(InputValue::new(
                "in",
                TypeRef::named_nn_list(filter_arg_type),
            ));

    let input_val_for_field = InputValue::new(
        field_name,
        TypeRef::named(complete_comparison_obj.type_name()),
    );

    input_objs.append(&mut vec![complex_comparison_obj, complete_comparison_obj]);
    (input_val_for_field, input_objs)
}
