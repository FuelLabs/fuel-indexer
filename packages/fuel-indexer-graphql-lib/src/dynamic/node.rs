//! `async_graphql::dynamic` extensions for handling GraphQL nodes.
//! See: https://graphql.org/learn/global-object-identification/#node-interface

use super::data::*;
use super::prelude::*;
use super::resolver::*;

pub type DynamicNodeLocalId = String;

pub type DynamicNodeTypeId = String;

#[derive(Clone, Hash, PartialEq, Eq, Debug)]
pub struct DynamicNodeId(DynamicNodeTypeId, DynamicNodeLocalId);
impl DynamicNodeId {
    pub fn new(
        type_id: impl Into<DynamicNodeTypeId>,
        local_id: impl Into<DynamicNodeLocalId>,
    ) -> Self {
        Self(type_id.into(), local_id.into())
    }
}
impl DynamicNodeId {
    pub fn type_id(&self) -> &DynamicNodeTypeId {
        &self.0
    }
    pub fn local_id(&self) -> &DynamicNodeLocalId {
        &self.1
    }
}
impl FromStr for DynamicNodeId {
    type Err = ();
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut split = s.splitn(2, ':');
        let type_id = split.next().unwrap().parse().unwrap();
        let local_id = split.next().unwrap().to_string();
        Ok(Self(type_id, local_id))
    }
}
impl From<DynamicNodeId> for String {
    fn from(id: DynamicNodeId) -> String {
        format!("{}:{}", id.type_id(), id.local_id())
    }
}

#[derive(Clone, PartialEq, Eq)]
pub struct DynamicNode(DynamicNodeTypeId, DynamicNodeLocalId, DynamicData);
impl DynamicNode {
    pub fn new(
        type_id: impl Into<DynamicNodeTypeId>,
        local_id: impl Into<DynamicNodeLocalId>,
        data: impl Into<DynamicData>,
    ) -> Self {
        Self(type_id.into(), local_id.into(), data.into())
    }

    pub fn id(&self) -> DynamicNodeId {
        DynamicNodeId::new(self.type_id(), self.local_id())
    }
    pub fn type_id(&self) -> &DynamicNodeTypeId {
        &self.0
    }
    pub fn local_id(&self) -> &DynamicNodeLocalId {
        &self.1
    }
    pub fn data(&self) -> &DynamicData {
        &self.2
    }
}

#[derive(Clone, Hash, PartialEq, Eq)]
pub struct DynamicConnectionField {
    pub id: DynamicFieldId,
    pub ref_node_type_id: DynamicNodeTypeId,
}

#[derive(Clone, Hash, PartialEq, Eq)]
pub struct DynamicRefField {
    pub id: DynamicFieldId,
    pub data_field_id: DynamicFieldId,
    pub ref_node_type_id: DynamicNodeTypeId,
}

#[derive(Clone, PartialEq, Eq)]
pub struct DynamicNodeType {
    pub id: DynamicNodeTypeId,
    pub data_fields: Vec<DynamicDataField>,
    pub ref_fields: Vec<DynamicRefField>,
    pub connection_fields: Vec<DynamicConnectionField>,
}

impl From<DynamicNodeType> for Object {
    fn from(value: DynamicNodeType) -> Self {
        let name = TypeRef::node(value.id.clone());

        let mut object = <Object as NodeObject<DynamicResolver>>::new_node(&name);

        // Add data fields
        for data_field in &value.data_fields {
            object = <Object as NodeObject<DynamicResolver>>::data_field(
                object,
                data_field.id(),
                TypeRef::named_nn(data_field.data_type().to_string()),
            );
        }

        // Add ref fields
        for ref_field in &value.ref_fields {
            object = <Object as NodeObject<DynamicResolver>>::ref_field(
                object,
                ref_field.id.clone(),
                TypeRef::named_nn(ref_field.ref_node_type_id.to_string()),
            );
        }

        // Add connection fields
        for connection_field in &value.connection_fields {
            object = <Object as NodeObject<DynamicResolver>>::connection_field(
                object,
                connection_field.id.clone(),
                TypeRef::named_nn(TypeRef::connection(
                    connection_field.ref_node_type_id.clone(),
                )),
            );
        }

        object
    }
}
