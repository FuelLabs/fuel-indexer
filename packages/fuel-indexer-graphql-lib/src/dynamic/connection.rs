//! `async_graphql::dynamic` extensions for handling GraphQL connections.
//! See: https://graphql.org/learn/pagination/#end-of-list-counts-and-connections
//! See: https://relay.dev/graphql/connections.htm#sec-Connection-Types

use super::node::*;
use super::paging::*;

#[derive(Clone, PartialEq, Eq)]
pub struct DynamicConnectionEdge {
    pub node_id: DynamicNodeId,
    pub cursor: DynamicCursor,
}

#[derive(Clone, PartialEq, Eq)]
pub struct DynamicConnection {
    pub total_count: usize,
    pub edges: Vec<DynamicConnectionEdge>,
    pub page_info: DynamicPageInfo,
}
