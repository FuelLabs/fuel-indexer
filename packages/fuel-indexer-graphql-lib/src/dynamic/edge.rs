//! `async_graphql::dynamic` extensions for handling GraphQL connections.
//! See: https://graphql.org/learn/pagination/#end-of-list-counts-and-connections
//! See: https://relay.dev/graphql/connections.htm#sec-Connection-Types

use super::node::*;
use super::prelude::*;

pub type DynamicEdgeTypeId = String;

#[derive(Clone, Hash, PartialEq, Eq)]
pub struct DynamicEdgeId(DynamicEdgeTypeId, DynamicNodeLocalId, DynamicNodeLocalId);
impl DynamicEdgeId {
    pub fn new(
        type_id: impl Into<DynamicEdgeTypeId>,
        tail_local_id: impl Into<DynamicNodeLocalId>,
        head_local_id: impl Into<DynamicNodeLocalId>,
    ) -> Self {
        Self(type_id.into(), tail_local_id.into(), head_local_id.into())
    }
}
impl From<DynamicEdgeId> for String {
    fn from(id: DynamicEdgeId) -> String {
        format!("{}:{}:{}", id.0, id.1, id.2)
    }
}

pub type DynamicEdgeData = serde_json::Value;

#[derive(Clone, PartialEq, Eq)]
pub struct DynamicEdge(
    DynamicEdgeTypeId,
    DynamicNodeLocalId,
    DynamicNodeLocalId,
    DynamicEdgeData,
);
impl DynamicEdge {
    pub fn new(
        id: impl Into<DynamicEdgeTypeId>,
        tail_local_id: impl Into<DynamicNodeLocalId>,
        head_local_id: impl Into<DynamicNodeLocalId>,
        data: impl Into<DynamicEdgeData>,
    ) -> Self {
        Self(
            id.into(),
            tail_local_id.into(),
            head_local_id.into(),
            data.into(),
        )
    }

    pub fn type_id(&self) -> &DynamicEdgeTypeId {
        &self.0
    }
    pub fn tail_local_id(&self) -> &DynamicNodeLocalId {
        &self.1
    }
    pub fn head_local_id(&self) -> &DynamicNodeLocalId {
        &self.2
    }
    pub fn data(&self) -> &DynamicEdgeData {
        &self.3
    }
}
