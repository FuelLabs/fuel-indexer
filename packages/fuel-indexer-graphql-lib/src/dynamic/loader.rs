use super::edge::*;
use super::node::*;
use super::prelude::*;

#[async_trait]
pub trait DynamicLoader: Send + Sync + 'static {
    async fn load_node_by_id(
        &self,
        id: &DynamicNodeId,
    ) -> Result<Option<DynamicNode>, ()>;
    async fn load_nodes_by_id(
        &self,
        ids: &[DynamicNodeId],
    ) -> Result<Vec<Option<DynamicNode>>, ()>;
    async fn load_edges(
        &self,
        type_id: &DynamicEdgeTypeId,
        tail_local_id: &DynamicNodeLocalId,
    ) -> Result<Vec<DynamicEdge>, ()>;
}

pub struct TestLoader {
    pub nodes: HashMap<DynamicNodeId, DynamicNode>,
    pub edges: HashMap<(DynamicEdgeTypeId, DynamicNodeLocalId), DynamicEdge>,
}
impl TestLoader {
    pub fn new(nodes: Vec<DynamicNode>, edges: Vec<DynamicEdge>) -> Self {
        let nodes = nodes
            .into_iter()
            .map(|node| (node.id(), node))
            .collect::<HashMap<_, _>>();
        let edges = edges
            .into_iter()
            .map(|edge| ((edge.type_id().clone(), edge.tail_local_id().clone()), edge))
            .collect::<HashMap<_, _>>();
        Self { nodes, edges }
    }
}

#[async_trait]
impl DynamicLoader for TestLoader {
    async fn load_node_by_id(
        &self,
        id: &DynamicNodeId,
    ) -> Result<Option<DynamicNode>, ()> {
        let node = self.nodes.get(id);
        Ok(node.cloned())
    }

    async fn load_nodes_by_id(
        &self,
        ids: &[DynamicNodeId],
    ) -> Result<Vec<Option<DynamicNode>>, ()> {
        let mut nodes = Vec::new();
        for id in ids {
            let node = self.load_node_by_id(id).await?;
            nodes.push(node);
        }
        Ok(nodes)
    }

    async fn load_edges(
        &self,
        type_id: &DynamicEdgeTypeId,
        tail_local_id: &DynamicNodeLocalId,
    ) -> Result<Vec<DynamicEdge>, ()> {
        let edges = self
            .edges
            .iter()
            .filter(|(id, _edge)| id.0 == *type_id && id.1 == *tail_local_id)
            .map(|(_, edge)| edge)
            .cloned()
            .collect();
        Ok(edges)
    }
}
