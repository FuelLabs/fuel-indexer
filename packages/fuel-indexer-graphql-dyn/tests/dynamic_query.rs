use assert_matches::*;
use async_graphql::dynamic::Schema;
use async_graphql::Request;
use async_graphql_value::ConstValue;
use fuel_indexer_graphql_dyn::testing::prelude::*;

use insta::*;

#[tokio::test]
async fn test() {
    let schema = Schema::build_test();
    let schema = schema.finish().unwrap();

    // Execute a query
    let response = schema
        .execute(Request::new(
            r#"
            query {
                transaction(id: "Transaction:0-0-0") {
                    id
                }
                chain(id: "Chain:0") {
                  transactions(first: 1) {
                      nodes {
                          id
                      }
                  }
                }
            }
        "#,
        ))
        .await
        .into_result()
        .unwrap();

    assert_json_snapshot!(response.data);

    assert_matches!(response.data, ConstValue::Object(_));
    if let ConstValue::Object(query_fields) = response.data {
        assert_eq!(query_fields.len(), 2);

        assert_matches!(query_fields.get("transaction"), Some(ConstValue::Object(_)));
        if let Some(ConstValue::Object(transaction_fields)) =
            query_fields.get("transaction")
        {
            assert_eq!(transaction_fields.len(), 1);
            let id_field = transaction_fields.get("id");
            assert_matches!(id_field, Some(ConstValue::String(_)));
            if let Some(ConstValue::String(id)) = id_field {
                assert_eq!(id, "Transaction:0-0-0");
            }
        }

        assert_matches!(query_fields.get("chain"), Some(ConstValue::Object(_)));
        if let Some(ConstValue::Object(chain_fields)) = query_fields.get("chain") {
            assert_eq!(chain_fields.len(), 1);
            let transactions_field = chain_fields.get("transactions");
            assert_matches!(transactions_field, Some(ConstValue::Object(_)));
            if let Some(ConstValue::Object(transactions_fields)) = transactions_field {
                assert_eq!(transactions_fields.len(), 1);
                let nodes_field = transactions_fields.get("nodes");
                assert_matches!(nodes_field, Some(ConstValue::List(_)));
                if let Some(ConstValue::List(nodes)) = nodes_field {
                    assert_eq!(nodes.len(), 1);
                    assert_matches!(nodes.get(0), Some(ConstValue::Object(_)));
                    if let Some(ConstValue::Object(node_fields)) = nodes.get(0) {
                        assert_eq!(node_fields.len(), 1);
                        let id_field = node_fields.get("id");
                        assert_matches!(id_field, Some(ConstValue::String(_)));
                        if let Some(ConstValue::String(id)) = id_field {
                            assert_eq!(id, "Transaction:0-0-0");
                        }
                    }
                }
            }
        }
    }
}
