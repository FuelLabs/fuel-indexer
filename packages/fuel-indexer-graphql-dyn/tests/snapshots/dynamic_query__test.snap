---
source: packages/fuel-indexer-graphql-dyn/tests/dynamic_query.rs
expression: response.data
---
{
  "transaction": {
    "id": "Transaction:0-0-0"
  },
  "chain": {
    "transactions": {
      "nodes": [
        {
          "id": "Transaction:0-0-0"
        }
      ]
    }
  }
}
